isotope/service/pkg/srv/handler.go

```diff
@@ -54,6 +57,15 @@ func (h Handler) ServeHTTP(writer http.ResponseWriter, request *http.Request) {
 		prometheus.RecordResponseSent(duration, len(h.responsePayload), status)
 	}
 
+	// Simulate failure based on the error percentage
+	reqCount := atomic.AddUint64(&h.counter, 1)
+	if h.Service.ErrorRate > 0 && reqCount >= (10000/uint64(h.Service.ErrorRate*10000)) {
```

Comment:
This seems like we have a race, multiple concurrent requests can enter the `if` before we clear it back out. Maybe we want to check if reqCount is _exactly_ `10000/uint64(h.Service.ErrorRate*10000` (maybe with some slightly altered math around the rounding/division/etc. Then if we get ErroRate=5 and reqCount=5 we enter; reqCount=6 we skip.

Although that does undercount errors since we will wipe it back to `0` on L63. So we either under count or over count...

On that note - why do we multiple and divide by 10000? isn't it the same?

Reply:
@howardjohn yes, it's definitely a race condition we likely cannot avoid without having a critical section around the full counter&counter reset logic - that'd be e.g. the original channel implementation or mutex.

In reality however, when simulating a 2% error rate, the drift over the course of 500k requests with ~5k rps on a single instance of isotope resulted in a peak of 43 requests off the expected failure count. That's a 0,4% drift which results in an actual failure rate of 0,200086%. IMHO this is negligible for a simpler in-code solution.
When simulating a way higher error rate, the drift becomes less noticeable. 


> On that note - why do we multiple and divide by 10000? isn't it the same?
You're 100% right. Just the uint64 cast would move, that works as well: 
```suggestion
	if h.Service.ErrorRate > 0 && reqCount >= uint64(1/h.Service.ErrorRate) {
```
You have a preference towards this solution? Then I'll open-up another PR

Reply:
One untested idea - just do `if increment() % ErrorRate {}`. And don't reset it. Then since there is only one operation we don't have any race. The only concern is if we hit the max int size, not sure what happens there - we likely just wrap around without issues